### DISCUSSION TITLE
Dependency to orjson breaks compatibility to PyPy

### DISCUSSION THREAD

**Discussion by rkrell** [fastapi/fastapi#11659] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
pip install fastapi==0.111.0
```

### Description

I use a virtual environment with pypy3.10-7.3.13.

Recently, there has been added a default dependency to orjson.
orson has declared to be incompatible to PyPy see https://pypi.org/project/orjson/:

> orjson does not and will not support PyPy.

When running the command
`pip install fastapi==0.111.0`

I end up in:

```
...
Building wheels for collected packages: orjson
  Building wheel for orjson (pyproject.toml) ... error
  error: subprocess-exited-with-error
  
  × Building wheel for orjson (pyproject.toml) did not run successfully.
  │ exit code: 1
  ╰─> [14 lines of output]
      Running `maturin pep517 build-wheel -i /home/rkrell/.pyenv/versions/pypy3.10-7.3.13/envs/myapi-pypy3.10-7.3.13/bin/python --compatibility off`
      📦 Including license file "/tmp/pip-install-gnxx4czm/orjson_a4021de98489486bb0484abd53cfbaa5/LICENSE-APACHE"
      📦 Including license file "/tmp/pip-install-gnxx4czm/orjson_a4021de98489486bb0484abd53cfbaa5/LICENSE-MIT"
      🍹 Building a mixed python/rust project
      🔗 Found pyo3-ffi bindings
      🐍 Found PyPy 3.10 at /home/rkrell/.pyenv/versions/pypy3.10-7.3.13/envs/myapi-pypy3.10-7.3.13/bin/python
      error: package `associative-cache v2.0.0` cannot be built because it requires rustc 1.65 or newer, while the currently active rustc version is 1.63.0
      Either upgrade to rustc 1.65 or newer, or use
      cargo update -p associative-cache@2.0.0 --precise ver
      where `ver` is the latest version of `associative-cache` supporting rustc 1.63.0
      💥 maturin failed
        Caused by: Failed to build a native library through cargo
        Caused by: Cargo build finished with "exit status: 101": `env -u CARGO PYO3_ENVIRONMENT_SIGNATURE="pypy-3.10-64bit" PYO3_PYTHON="/home/rkrell/.pyenv/versions/pypy3.10-7.3.13/envs/myapi-pypy3.10-7.3.13/bin/python" PYTHON_SYS_EXECUTABLE="/home/rkrell/.pyenv/versions/pypy3.10-7.3.13/envs/myapi-pypy3.10-7.3.13/bin/python" "cargo" "rustc" "--message-format" "json-render-diagnostics" "--manifest-path" "/tmp/pip-install-gnxx4czm/orjson_a4021de98489486bb0484abd53cfbaa5/Cargo.toml" "--release" "--lib"`
      Error: command ['maturin', 'pep517', 'build-wheel', '-i', '/home/rkrell/.pyenv/versions/pypy3.10-7.3.13/envs/myapi-pypy3.10-7.3.13/bin/python', '--compatibility', 'off'] returned non-zero exit status 1
      [end of output]
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
  ERROR: Failed building wheel for orjson
Failed to build orjson
ERROR: Could not build wheels for orjson, which is required to install pyproject.toml-based projects
```

Is fastapi now officially declared to be incompatible to PyPy (although it has been working the whole time) or is there a way to exclude orjson from the requirements for this purpose?

### Operating System

Linux

### Operating System Details

Debian 12

### FastAPI Version

0.111.0

### Pydantic Version

2.7.2

### Python Version

Python 3.10.13 (f1607341da97ff5a1e93430b6e8c4af0ad1aa019, Sep 28 2023, 05:41:26) [PyPy 7.3.13 with GCC 10.2.1 20210130 (Red Hat 10.2.1-11)]

**Comment by Kludex** [comment, marked as answer]
An environment marker is needed for PyPy: `orjson; platform_python_implementation == 'CPython'`.

**Reply by rkrell** [reply]
Yes, but probably in FastAPI itself. Currently I cannot see any way to prevent `orjson` from being installed when updating to the latest version of `fastapi`.

**Reply by Kludex** [reply]
> Yes, but probably in FastAPI itself. Currently I cannot see any way to prevent `orjson` from being installed when updating to the latest version of `fastapi`.

Yes, I meant fastapi itself.

**Reply by danyd555** [reply]
is there already an issue for this ? 
only workaround is playing with pip install --no-deps and pip check to install only the packages you want,

**Comment by rkrell** [comment]
I created a pull request against master for this: #11689
Some checks fail, but this PR should not be the cause as it seems.

I changed the specifier to `platform_python_implementation != 'PyPy'`, to be more specific to this issue and to not exclude other python implementations for now.

**Reply by tiangolo** [reply]
Thank you!

Given this and another report, I decided to remove `orjson` and `ujson` from the default dependencies completely. They are still in the `all` optional dependencies. But they are no longer included when installing with `pip install fastapi`.

This is available in FastAPI 0.111.1.

Thanks for letting me know and for the PR! :coffee:
